nni/algorithms/compression/v2/pytorch/pruning/tools/base.py

```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
+        else:
+            return None
+
+    def _expand_mask(self, module_name: str, target_name: str, mask: Tensor) -> Tensor:
+        # Expand the shrinked mask to the pruning target size.
+        scalor = self._get_scalor(module_name=module_name, target_name=target_name)
+        if scalor:
+            wrapper = self.pruner.get_modules_wrapper()[module_name]
+            return scalor.expand(mask, getattr(wrapper, f'{target_name}_mask').shape)
+        else:
+            return mask.clone()
+
+    def _shrink_mask(self, module_name: str, target_name: str, mask: Tensor) -> Tensor:
+        # Shrink the mask by scalor, shrinked mask usually has the same size with metric.
+        scalor = self._get_scalor(module_name=module_name, target_name=target_name)
+        if scalor:
+            mask = (scalor.shrink(mask) != 0).type_as(mask)
+        return mask
+
+    def _continuous_mask(self, new_masks: Dict[str, Dict[str, Tensor]]) -> Dict[str, Dict[str, Tensor]]:
+        # Set the already masked part in the metric to the minimum value.
+        target_name = 'weight'
+        for module_name, target_mask in new_masks.items():
+            wrapper = self.pruner.get_modules_wrapper()[module_name]
+            old_target_mask = getattr(wrapper, f'{target_name}_mask', None)
+            if old_target_mask is not None:
+                new_masks[module_name][target_name] = torch.min(target_mask[target_name], old_target_mask)
+        return new_masks
+
+    def common_target_masks_generation(self, metrics: Dict[str, Tensor]) -> Dict[str, Dict[str, Tensor]]:
         """
+        Generate masks for metrics-dependent targets.
+
         Parameters
         ----------
         metrics
-            A metric dict. The key is the name of layer, the value is its metric.
+            The format is {module_name: weight_metric}.
+            The metric of `weight` usually has the same size with shrinked mask.
+
+        Return
+        ------
+        Dict[str, Dict[str, Tensor]]
+            The format is {module_name: {target_name: mask}}.
+            Return the masks of the same size as its target.
         """
         raise NotImplementedError()
 
-    def _expand_mask(self, name: str, mask: Tensor) -> Dict[str, Tensor]:
+    def special_target_masks_generation(self, masks: Dict[str, Dict[str, Tensor]]) -> Dict[str, Dict[str, Tensor]]:
         """
+        Some pruning targets' mask generation depends on other targets, i.e., bias mask depends on weight mask.
+        This function is used to generate these masks, and it be called at the end of `generate_sparsity`.
+
         Parameters
         ----------
-        name
-            The masked module name.
-        mask
-            The reduced mask with `self.dim` and `self.block_sparse_size`.
-
-        Returns
-        -------
-        Dict[str, Tensor]
-            The key is `weight` or `bias`, value is the final mask.
+        masks
+            The format is {module_name: {target_name: mask}}.
```

Comment:
I noticed "{module_name: {target_name: mask}}" in many places, is it better to define a data structure for it?

Reply:
good suggestion, will add one, but maybe `{module_name: {target_name: mask}}` maybe is more friendly for review.